### PR TITLE
refactor(permissions): factor out PermissionPrompter trait, add callbacks

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit refactors several things in "runtime/permissions" module:
- splits it into "mod.rs" and "prompter.rs"
- adds "PermissionPrompter" trait with two implementations:
 * "TtyPrompter"
 * "TestPrompter"
- adds "before" and "after" prompt callback which can be used to hide
progress bar in the CLI (to be done in a follow up)
- "permissions_prompt" API returns "PromptResponse" enum, instead
of a boolean; this allows to add "allow all"/"deny all" functionality for
the prompt

Towards https://github.com/denoland/deno/issues/16860

### CODE REVIEW

File: runtime/permissions.rs [hunk 1]
```diff
@@ -24,16 +23,216 @@ use std::hash::Hash;
...
-use std::sync::atomic::AtomicBool;
-#[cfg(test)]
-use std::sync::atomic::Ordering;
 
 const PERMISSION_EMOJI: &str = "⚠️";
 
 static DEBUG_LOG_ENABLED: Lazy<bool> =
   Lazy::new(|| log::log_enabled!(log::Level::Debug));
 
+// TODO(bartlomieju): make a Prompter struct (or anything that implements
```
Comment (bartlomieju) [inline runtime/permissions.rs:32]: TODO: split this module into multiple files

File: runtime/permissions/prompter.rs [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+pub trait PermissionPrompter: Send + Sync {
+  fn prompt(
+    &mut self,
+    message: &str,
+    name: &str,
+    api_name: Option<&str>,
+  ) -> bool;
+  fn set_before_prompt_callback(&mut self, callback: Option<PromptCallback>);
+  fn set_after_prompt_callback(&mut self, callback: Option<PromptCallback>);
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply runtime/permissions/prompter.rs:45]: Err, not sure I get it, can you elaborate?

**Comment by bartlomieju** [comment]
@dsherret please take a look

**Review by dsherret — APPROVED** [review]
LGTM!

I was thinking last night how we can integrate this with the existing progress bar. I'm thinking we should tackle it in a patch after the minor release.

**Comment by bartlomieju** [comment]
> LGTM!
> 
> I was thinking last night how we can integrate this with the existing progress bar. I'm thinking we should tackle it in a patch after the minor release.

Then I won't rush it, let's just do it in this PR after the release

**Comment by bartlomieju** [comment]
Note to self: refactor `permission_prompt` function to return an enum, this will allow us to add a prompt that is not binary, but enables to allow all permissions for certain flag (or deny them)

**Pull request merged**